### PR TITLE
[API Implementation]: System.Diagnostics.CodeAnalysis.StringSyntaxAttribute

### PR DESCRIPTION
**Pull request by deeprobin**
Proposal implementation of #62505 

# Proposal 
```cs
namespace System.Diagnostics.CodeAnalysis
{
    [AttributeUsage(AttributeTargets.Parameter |
                    AttributeTargets.Field |
                    AttributeTargets.Property,
                    AllowMultiple = false,
                    Inherited = false)]
    public sealed class StringSyntaxAttribute : Attribute
    {
        public StringSyntaxAttribute(string syntax);
        public string Syntax { get; }

        public const string Regex = "regex";

        // As we're adding more support we can add new languages like:
        // public const string Xml = "xml";
        // public const string Json = "json";
    }
}
```

/cc @stephentoub

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/StringSyntaxAttribute.cs
Comment (danmoseley) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/StringSyntaxAttribute.cs:29]: looking at our other attributes with constructors taking strings, there seems to be a predominant pattern of not validating it for null/empty. so this is fine.

File: src/libraries/System.Runtime/ref/System.Runtime.cs
Comment (danmoseley) [inline src/libraries/System.Runtime/ref/System.Runtime.cs:9178]: some tabs crept in here. could you please instead do ref file updates with the tool?
https://github.com/dotnet/runtime/blob/2879e6ca886eb084335ccd0a579019fbe359628e/docs/coding-guidelines/updating-ref-source.md

File: src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems
Comment (danmoseley) [inline src/libraries/System.Private.CoreLib/src/System.Private.CoreLib.Shared.projitems:61]: very nit: if you're touching it anyway you could remove the `\` before `Internal` and sort those lines above.

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/StringSyntaxAttribute.cs
Comment (danmoseley) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/StringSyntaxAttribute.cs:7]:
```suggestion
    /// <see cref="StringSyntaxAttribute"/> tells an IDE the type of content that a string contains. It can use this to colorize or validate the contents of the string.
```
?

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/StringSyntaxAttribute.cs
Comment (danmoseley) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/StringSyntaxAttribute.cs:29]: also the compiler will validate for null.

**Comment by danmoseley** [comment]
LGTM with the small notes above. thanks

**Comment by danmoseley** [comment]
The regex test crash is obviously unrelated and should not block merge. Looking into it offline.

File: src/libraries/System.Runtime/ref/System.Runtime.cs
Comment (deeprobin) [reply src/libraries/System.Runtime/ref/System.Runtime.cs:9178]: The tool does not work for me :/
 
`dotnet build --no-incremental /t:GenerateReferenceSource`
grafik

Comment (deeprobin) [reply src/libraries/System.Runtime/ref/System.Runtime.cs:9178]: I indented it manually

Comment (danmoseley) [reply src/libraries/System.Runtime/ref/System.Runtime.cs:9178]: hmm, the error is cut off so not sure what that is, a binary mismatch most likely. I wonder whether it would work if you used the dotnet.cmd in the src folder.

**Comment by danmoseley** [comment]
Oh nuts, I forgot. Could you please add some simple test? You can see we do that even for trivial attribute types, eg ConditionalAttributeTests.cs, RequiresUnreferencedCodeAttributeTests.cs

**Comment by stephentoub** [comment]
The work here also includes attributing the parameters that need to be attributed.

**Comment by danmoseley** [comment]
> The work here also includes attributing the parameters that need to be attributed.

Do you mean all the places on Regex? (Which I assume are already hard coded in VS)

Outside of Regex, since we're not using RegexGenerator ourselves yet, I only found this one place (unless we annotate tests)
```c#
C:\git\runtime\src\libraries\System.Configuration.ConfigurationManager\ref\System.Configuration.ConfigurationManager.cs:
  934      public partial class RegexStringValidator : System.Configuration.ConfigurationValidatorBase
  935      {
  936:         public RegexStringValidator(string regex) { }
  937          public override bool CanValidate(System.Type type) { throw null; }
  938          public override void Validate(object value) { }
  ...
  941      public sealed partial class RegexStringValidatorAttribute : System.Configuration.ConfigurationValidatorAttribute
  942      {
  943:         public RegexStringValidatorAttribute(string regex) { }
  944          public string Regex { get { throw null; } }
  945          public override System.Configuration.ConfigurationValidatorBase ValidatorInstance { get { throw null; } }
  ```

**Comment by danmoseley** [comment]
Oh there's a couple in System.Security.Permissions that aren't worth touching. I assume you don't mean annotating our implementations.

Oh, and of course the ref for RegexGeneratorAttribute, since VS doesn't know about that yet - the original example.

**Comment by stephentoub** [comment]
> I assume you don't mean annotating our implementations.

We should annotate both the src and ref for anything being annotated. There's no reason to introduce ref-only annotations here.

> Do you mean all the places on Regex?

At a minimum on Regex and RegexGenerator.

I also think we should be adding Json and Xml and annotating for those as part of this.  It doesn't matter that VS doesn't yet light-up with them... VS isn't the only possible consumer here,  and with an attribute for saying "this string is in this well-established language", we should consistently attribute things that can self-identify. Subsequently we could annotate more,  like Html, Uri, etc.

**Comment by stephentoub** [comment]
> Oh there's a couple in System.Security.Permissions

There's also WebProxy's bypassList ctor parameter. It's an array of regexes rather than a single one,  but presumablywe should consider it acceptable to apply this to a collection of strings and not just an individual one, in particular if that collection can be constructed with collection-initialization syntaxes

**Comment by stephentoub** [comment]
There's also RegularExpressionAttribute.

**Comment by danmoseley** [comment]
OK, but let's limit this PR to regex, not json xml or anything else.

I'm also fine with merging this (with test) and doing a follow up PR to add the usage.

**Comment by stephentoub** [comment]
> I'm also fine with merging this (with test) and doing a follow up PR to add the usage.

I strongly prefer us to use new APIs in the PR that adds them. It helps validate the API, and frequently reveals flaws in the approved design,  even for seemingly simple cases. This is about as simple as they come and so we may not hit any stumbling blocks, but I still like the precedent.

**Comment by stephentoub** [comment]
> OK, but let's limit this PR to regex, not json xml or anything else.

You'll open an issue to track doing that then?

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/StringSyntaxAttribute.cs
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/StringSyntaxAttribute.cs:7]: This isn't just for an IDE. It's for any static code analysis.

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/StringSyntaxAttribute.cs
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/StringSyntaxAttribute.cs:24]: I don't understand what this is saying.

**Comment by danmoseley** [comment]
> I strongly prefer us to use new APIs in the PR that adds them. It helps validate the API, and frequently reveals flaws in the approved design,  even for seemingly simple cases.

Ok

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/StringSyntaxAttribute.cs
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/StringSyntaxAttribute.cs:17]: We haven't been adding code examples into xml comments like this.  @carlossanlop, is this the recommended way moving forward?

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/StringSyntaxAttribute.cs
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/StringSyntaxAttribute.cs:35]:
```suggestion
        /// Gets the syntax identifier for strings containing regular expressions. 
```

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/StringSyntaxAttribute.cs
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/StringSyntaxAttribute.cs:20]: This was approved as AllowMultiple = false, Inherited = false. The default for Inherited is true.

Comment (deeprobin) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/StringSyntaxAttribute.cs:20]: Yes, but the Inherited is unnessecary on non-method or classes.

For reference VS22 message:
grafik

& AllowMultiple = false is the default

**Comment by deeprobin** [comment]
> Oh nuts, I forgot. Could you please add some simple test? You can see we do that even for trivial attribute types, eg ConditionalAttributeTests.cs, RequiresUnreferencedCodeAttributeTests.cs

@danmoseley 
I implemented a simple test case like in other attribute tests.
https://github.com/dotnet/runtime/pull/62995/files#diff-fb2b9ba564351f9e52979d1853efb4c9f9a639e1dff99f1b9d72116541fe7bfdR10-R16

I chose a Theory for this to add other syntax values as InlineData in future development (see https://github.com/dotnet/runtime/pull/62995#issuecomment-999225316).

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/StringSyntaxAttribute.cs
Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/StringSyntaxAttribute.cs:20]: > For reference VS22 message:

What is producing that message?  What's the diagnostic code?

Comment (deeprobin) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/StringSyntaxAttribute.cs:20]: I think its the code analyzer of `AttributeUsageAttribute`. What the diagnostics code is in the background I unfortunately can't tell you.

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/StringSyntaxAttribute.cs
Comment (danmoseley) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/StringSyntaxAttribute.cs:17]: the guidance he has in https://github.com/carlossanlop/DocsPortingTool/blob/main/BackportInstructions.md indicates all snippets should be in a file. @carlossanlop is that indeed the case for the shortest ones like this?

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/StringSyntaxAttribute.cs
Comment (danmoseley) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/StringSyntaxAttribute.cs:20]: The analyzer is faulty because IsInherited makes sense with AttributeTargets.Property as well. 

This does not seem to be an analyzer in dotnet/roslyn-analyzers. Do you have any Visual Studio extensions installed, that add analyzers? 

You can increase the verbosity of the build in Tools>Options>Build and then look at the build info in the Output Window and get the command line. That should include whatever this analyzer is.

Meanwhile you can ignore it/ disable this warning.

**Review by danmoseley — APPROVED** [review]

**Review by danmoseley — DISMISSED** [review]
Didn't mean to mark approved. Please add necessary use of this attribute, as requested above by @stephentoub.

**Comment by stephentoub** [comment]
We also shouldn't merge this into this conversation is sorted out:
https://github.com/dotnet/runtime/issues/62505#issuecomment-1006262593

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/StringSyntaxAttribute.cs
Comment (deeprobin) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/StringSyntaxAttribute.cs]: I think that was because of R#. I have now adopted the part as in the approval.

**Comment by deeprobin** [comment]
Can somebody re-run CI? API-Compat does not like that I changed the AttributeUsage...

**Comment by stephentoub** [comment]
> Can somebody re-run CI? API-Compat does not like that I changed the AttributeUsage...

Restarting it won't help. The problem is in the PR. The src and ref declarations differ.

**Comment by stephentoub** [comment]
I added a commit that fixes the API to match the revamped shape and uses the attribute in appropriate places. @CyrusNajmabadi, please take a look at the usage in particular to make sure it aligns with your expectations. Thanks.

File: src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/StringSyntaxAttribute.cs
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/StringSyntaxAttribute.cs:44]: @terrajobst, @bartonjs, @CyrusNajmabadi, the issue described adding "DateTimeFormat", PascalCased.  We previously had "regex", lower-cased.  I made them both PascalCased (and it looks like VS' detection of "lang=regex" already supports "Regex".  Please let me know if we should go back to having different casings or if we'd prefer for a different consistent style.

Comment (bartonjs) [reply src/libraries/System.Private.CoreLib/src/System/Diagnostics/CodeAnalysis/StringSyntaxAttribute.cs:44]: Personally, I like  `public const string SomethingHere = nameof(SomethingHere);` as a followable pattern.

**Comment by CyrusNajmabadi** [comment]
I would very much prefer PascalCased

File: src/libraries/Microsoft.Extensions.Logging.Console/ref/Microsoft.Extensions.Logging.Console.cs
Comment (stephentoub) [inline src/libraries/Microsoft.Extensions.Logging.Console/ref/Microsoft.Extensions.Logging.Console.cs:44]: FYI, @dotnet/area-extensions-logging

File: src/libraries/Microsoft.Extensions.Logging.Console/ref/Microsoft.Extensions.Logging.Console.csproj
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Logging.Console/ref/Microsoft.Extensions.Logging.Console.csproj:17]: I think this would be more version-resilient if we wrote it like this:

```xml
<ItemGroup Condition="!$([MSBuild]::IsTargetFrameworkCompatible('$(TargetFramework)', 'net7.0'))"
```

That way when `$(NetCoreAppCurrent)` == `net8.0` and `$(NetCoreAppMinimum)` == `net7.0`, we don't need to update this condition.

Comment (stephentoub) [reply src/libraries/Microsoft.Extensions.Logging.Console/ref/Microsoft.Extensions.Logging.Console.csproj:17]: Thanks

**Review by stephentoub — APPROVED** [review]

**Comment by danmoseley** [comment]
@stephentoub did we need an issue to annotate for XML and JSON?

**Comment by stephentoub** [comment]
> did we need an issue to annotate for XML and JSON?

Only if we know someone will consume the annotations.  No one jumped at my suggestions around that.  If we want to be our own consumer and write an analyzer for json, ala #64009, or if someone in Roslyn is interested in writing a colorizer or the like for JSON, then yes.  Otherwise, it's not clear the annotations will be meaningful.  Bit of a catch-22 I guess.

**Comment by danmoseley** [comment]
@CyrusNajmabadi thoughts on those two? (Perhaps I missed your thoughts earlier). If there's a chance tooling will use in future, we can add now.

**Comment by CyrusNajmabadi** [comment]
Note: We have JSON support here https://github.com/dotnet/roslyn/pull/42873

We just haven't merged it.  If you guys added this, that would more strongly motivate things.  @jinujoseph  for thoughts here.

**Comment by danmoseley** [comment]
Let's add it. I hear JSON is quite popular these days. 😃

**Comment by stephentoub** [comment]
> Let's add it. I hear JSON is quite popular these days. 😃

https://github.com/dotnet/runtime/pull/64081

File: src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleFormatterOptions.cs
Comment (CyrusNajmabadi) [inline src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleFormatterOptions.cs:26]: oh interesting.  so we're doing this for more than just arguments.  that's def something i didn't realize.  will have to update roslyn side for sure to understand that :)

Comment (stephentoub) [reply src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleFormatterOptions.cs:26]: You and Jason pushed for the attribute to be for properties and fields, too. So, yeah :-)

**Pull request merged**
